spring-xd-dirt/src/main/java/org/springframework/xd/dirt/module/SynchronizingModuleRegistry.java

```diff
@@ -0,0 +1,151 @@
+/*
+ *
+ *  * Copyright 2011-2014 the original author or authors.
+ *  *
+ *  * Licensed under the Apache License, Version 2.0 (the "License");
+ *  * you may not use this file except in compliance with the License.
+ *  * You may obtain a copy of the License at
+ *  *
+ *  *      http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the License is distributed on an "AS IS" BASIS,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package org.springframework.xd.dirt.module;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.springframework.core.io.support.PathMatchingResourcePatternResolver;
+import org.springframework.core.io.support.ResourcePatternResolver;
+import org.springframework.util.Assert;
+import org.springframework.xd.dirt.core.RuntimeIOException;
+import org.springframework.xd.module.ModuleDefinition;
+import org.springframework.xd.module.ModuleType;
+import org.springframework.xd.module.SimpleModuleDefinition;
+
+/**
+ * A ModuleRegistry that is configured with two delegates: a source and a target registry. This registry will return
+ * results that exist in the source registry but will synchronize them with the target registry. Results returned are always
+ * from the <em>target</em> registry (unless composed). Mutative operations go through to the source repository though.
+ *
+ * <p>This is useful as reading
+ * Boot uber-jars requires local {@code java.io.File} access, but modules may reside in a remote registry (<i>e.g.</i>
+ * backed by HDFS). For such a case, simply use this registry as the main registry, configuring it with the {@code hdfs://} registry as
+ * its source and the {@code file://} one as its target.</p>
+ *
+ * @author Eric Bottard
+ */
+public class SynchronizingModuleRegistry implements WritableModuleRegistry {
+
+	private final WritableModuleRegistry sourceRegistry;
+
+	private final WritableModuleRegistry targetRegistry;
+
```

Comment:
Yeah, I've been using local and remote in some places. But note that they could both be "local" (e.g. different paths of the same (main) filesystem, as the unit tests do).